module/block_iterator.go

```diff
@@ -0,0 +1,114 @@
+package module
+
+import (
+	"fmt"
+
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// IterateJob defines the range of blocks to iterate over
+// the range could be either view based range or height based range.
+// when specifying the range, the start and end are inclusive, and the end must be greater than or
+// equal to the start
+type IterateJob struct {
+	Start uint64 // the start of the range
+	End   uint64 // the end of the range
+}
+
+// IterateJobCreator is an interface for creating iterate jobs
+type IteratorJobCreator interface {
+	// CreateJob takes a progress reader which is used to read the progress of the iterator
+	// and returns an iterate job that specifies the range of blocks to iterate over
+	CreateJob(IterateProgressReader) (IterateJob, error)
+}
+
+// IterateProgressReader reads the progress of the iterator, useful for resuming the iteration
+// after restart
+type IterateProgressReader interface {
+	// ReadNext reads the next block to iterate
+	// caller must ensure the reader is created by the IterateProgressInitializer,
+	// otherwise ReadNext would return exception.
+	ReadNext() (uint64, error)
+}
+
+// IterateProgressWriter saves the progress of the iterator
+type IterateProgressWriter interface {
+	// SaveNext persists the next block to be iterated
+	SaveNext(uint64) error
+}
+
+// IterateProgressInitializer is an interface for initializing the progress of the iterator
+// a initializer must be used to ensures the initial next block to be iterated is saved in
+// storage before creating the block iterator
+type IterateProgressInitializer interface {
+	Init() (IterateProgressReader, IterateProgressWriter, error)
+}
+
+// BlockIterator is an interface for iterating over blocks
+type BlockIterator interface {
```

Comment:
The BlockIterator interface can be implemented into height based iterator and view based iterator. 

The block iterator not long can be used by chunk data pack pruner, but alsoin future to implement protocol state pruner. 

The height based iterator is easy to implement, however, it can't guarantee to prune all data, since it doesn't iterate unfinalized blocks. The view based iterator can guarantee all blocks are pruned, but it's more complicated to implement. 

In this PR, I first implement the height based iterator, for chunk data pack, it's OK that we only prune by height, however, for protocol state, it's better that we can prune by view and ensure a more throughout pruning.